File: assessment.pl
#!/usr/bin/perl
use strict;
use warnings;
use feature 'say';

sub perl_is_awesome {
    # Implicit variable scoping
    $implicit_var = 10;
    say "Implicit variable: $implicit_var";
    {
        $implicit_var = "Now I'm a string!";
        say "Implicit variable changed: $implicit_var";
    }
    say "Implicit variable outside block: $implicit_var";

    # Type coercion issues
    my $num = 10;
    my $str = "5";
    my $coerced_result = $num + $str; # Numeric context
    say "Coerced result (numeric context): $coerced_result";
    $coerced_result = $num . $str; # String context
    say "Coerced result (string context): $coerced_result";

    # Context-sensitive behavior
    my @array = (1, 2, 3);
    my $scalar = @array; # Scalar context
    say "Array in scalar context: $scalar";
    my @new_array = @array; # List context
    say "Array in list context: @new_array";

    # Regex pitfalls
    my $text = "Perl is awesome!";
    if ($text =~ /awesome/) {
        say "Regex matched!";
    }
    my $number = "1234";
    if ($number =~ /(\d+)/) {
        say "Captured number: $1";
    }

    # Signal handling issues
    $SIG{INT} = sub {
        say "Caught SIGINT signal";
        exit 1;
    };

    # Use of eval
    my $code = 'say "Eval code executed!"';
    eval $code;
    if ($@) {
        say "Eval failed: $@";
    }

    # Bareword filehandles
    open(FH, '>', 'output.txt') or die "Could not open file: $!";
    print FH "Writing to a file using bareword filehandle\n";
    close(FH);

    # Implicit returns
    say "Implicit return example: ", implicit_return();

    # Variable shadowing
    my $var = 42;
    {
        my $var = "Shadowed variable";
        say "Inside block: $var";
    }
    say "Outside block: $var";

    # Autovivification pitfalls
    my %hash;
    $hash{foo}{bar} = 1;
    say "Autovivified hash: ", $hash{foo}{bar};

    # Using the `goto` statement
    my $count = 3;
    say "Using goto statement:";
    COUNTDOWN: while ($count) {
        say $count--;
        goto COUNTDOWN if $count > 0;
    }

    # JAPH example (Just Another Perl Hacker)
    $_='987;s/^(\d{1,3})/chr $1+114/e;eval';eval;print

    # Obfuscated code
    my $obfuscated = q{
        sub O{my $o=shift; $o=~s/(.)(.*)/$2$1/;return$o;} sub S{my $s=shift;
        $s=~tr/a-zA-Z/n-za-mN-ZA-M/;return$s;} sub U{my $u=shift; $u=~s/^(.)/$1$1/;return$u;}
        my $x="rtebl az gn!";
        for my $fn (qw(O S U)) { $x = $fn->($x); }
        say "Obfuscated code: $x";
    };
    eval $obfuscated;

    # Misusing globals and dynamic variables
    local $/;  # Slurp mode
    open my $fh, '<', $0 or die $!;
    my $code = <$fh>;
    close $fh;
    say "Read entire script using global variable \$/: $code";

    # Unintended list flattening
    my @list1 = (1, 2, 3);
    my @list2 = (4, 5, 6);
    my @flattened = (@list1, @list2);
    say "Flattened list: @flattened";

    # Obscure regular expression
    my $obscure_re = 's/([^\W_])([^\W_]*)/$2$1ay/g';
    my $obscure_text = "Pig Latin is fun";
    eval "\$obscure_text =~ $obscure_re";
    say "Obscure regex: $obscure_text";

    # Variable aliasing with typeglobs
    *foo = \$var;
    $foo = "Variable aliasing with typeglobs";
    say "Variable aliasing: $var";

    say "End of wildly awesome function.";
}

sub implicit_return {
    my $value = "Implicit return value";
    $value; # No explicit return statement
}

perl_is_awesome();

File: assessment.rb
#!/usr/bin/env ruby

def ruby_is_awesome
    # Dynamic typing issues
    dynamic_var = 10
    puts "Dynamic variable: #{dynamic_var}"
    dynamic_var = "Now I'm a string!"
    puts "Dynamic variable changed: #{dynamic_var}"
  
    # Variable shadowing
    var = 42
    1.times do
      var = "Shadowed variable"
      puts "Inside block: #{var}"
    end
    puts "Outside block: #{var}"
  
    # Type coercion issues
    num = 10
    str = "5"
    coerced_result = num + str.to_i
    puts "Coerced result (numeric context): #{coerced_result}"
    coerced_result = num.to_s + str
    puts "Coerced result (string context): #{coerced_result}"
  
    # Method overloading pitfalls
    def overloaded_method(x)
      if x.is_a?(Integer)
        "Integer method"
      elsif x.is_a?(String)
        "String method"
      else
        "Other method"
      end
    end
    puts "Overloaded method with Integer: #{overloaded_method(10)}"
    puts "Overloaded method with String: #{overloaded_method('hello')}"
    puts "Overloaded method with Array: #{overloaded_method([1, 2, 3])}"
  
    # Implicit returns
    def implicit_return
      value = "Implicit return value"
      value # No explicit return statement
    end
    puts "Implicit return example: #{implicit_return}"
  
    # Metaprogramming issues
    class MetaClass
      def method_missing(name, *args)
        puts "Called method: #{name} with args: #{args}"
      end
    end
    meta_instance = MetaClass.new
    meta_instance.non_existent_method("arg1", "arg2")
  
    # Monkey patching pitfalls
    class String
      def reverse
        "Monkey patched reverse"
      end
    end
    puts "Monkey patched String reverse: #{"hello".reverse}"
  
    # Symbol vs. String issues
    hash = { symbol_key: "value", "string_key" => "value" }
    puts "Symbol key: #{hash[:symbol_key]}"
    puts "String key: #{hash['string_key']}"
  
    # Block scope issues
    arr = [1, 2, 3]
    sum = 0
    arr.each { |sum| sum += 1 }
    puts "Block scope issue: sum = #{sum}"
  
    # Enumerable method pitfalls
    even_numbers = (1..10).select(&:even?)
    puts "Even numbers: #{even_numbers}"
  
    # Threading issues
    counter = 0
    threads = 10.times.map do
      Thread.new do
        1000.times { counter += 1 }
      end
    end
    threads.each(&:join)
    puts "Threading issues with counter: #{counter}"
  
    # File handling pitfalls
    begin
      file = File.open("nonexistent_file.txt")
    rescue => e
      puts "Caught file handling error: #{e.message}"
    ensure
      file.close if file
    end
  
    # Eval pitfalls
    code = 'puts "Eval code executed!"'
    eval(code)
  
    # Unintended list flattening
    nested_array = [1, [2, [3]], 4]
    flattened_array = nested_array.flatten
    puts "Flattened array: #{flattened_array}"
  
    # Regular expression pitfalls
    text = "Ruby is awesome!"
    if text =~ /awesome/
      puts "Regex matched!"
    end
    number = "1234"
    if number =~ /(\d+)/
      puts "Captured number: #{$1}"
    end
  
    # Unreadable code using DSLs
    unreadable_code = Class.new do
      def method_missing(name, *args, &block)
        puts "#{name}(#{args.map(&:inspect).join(', ')})"
        block.call if block
      end
    end.new
    unreadable_code.foo.bar.baz { unreadable_code.qux }
  
    # Complex metaprogramming issues
    class ComplexMeta
      def self.create_method(name)
        define_method(name) do |*args|
          puts "Called method: #{name} with args: #{args}"
        end
      end
    end
    ComplexMeta.create_method(:dynamic_method)
    ComplexMeta.new.dynamic_method("arg1", "arg2")
  
    # Performance issues with dynamic features
    start_time = Time.now
    1_000_000.times do
      dynamic_method = "method_#{rand(100)}"
      ComplexMeta.new.send(dynamic_method, "arg") rescue nil
    end
    puts "Performance issue time: #{Time.now - start_time} seconds"
  
    # Version compatibility problems
    if RUBY_VERSION < '2.5'
      puts "This feature requires Ruby 2.5 or higher"
    end
  
    # Memory leaks due to symbol creation
    10_000.times do |i|
      "string#{i}".to_sym
    end
    puts "Created many symbols, check memory usage"
  
    # Implicit conversion pitfalls
    def implicit_conversion(x, y)
      x + y
    end
    puts "Implicit conversion: #{implicit_conversion(10, "5".to_i)}"
    begin
      puts "Implicit conversion error: #{implicit_conversion(10, "five")}"
    rescue => e
      puts "Caught implicit conversion error: #{e.message}"
    end
  
    # Confusing inheritance
    class Base
      def initialize
        @value = "Base"
      end
  
      def show_value
        puts @value
      end
    end
  
    class Derived < Base
      def initialize
        @value = "Derived"
      end
    end
  
    Derived.new.show_value
  
    puts "End of wildly awesome function."
  end
  
  ruby_is_awesome
  